helix-core/src/main/java/org/apache/helix/util/WeightAwareRebalanceUtil.java

```diff
@@ -0,0 +1,141 @@
+package org.apache.helix.util;
+
+import org.apache.helix.HelixException;
+import org.apache.helix.ZNRecord;
+import org.apache.helix.api.config.RebalanceConfig;
+import org.apache.helix.api.rebalancer.constraint.AbstractRebalanceHardConstraint;
+import org.apache.helix.api.rebalancer.constraint.AbstractRebalanceSoftConstraint;
+import org.apache.helix.controller.common.PartitionStateMap;
+import org.apache.helix.controller.common.ResourcesStateMap;
+import org.apache.helix.controller.rebalancer.strategy.ConstraintRebalanceStrategy;
+import org.apache.helix.controller.stages.ClusterDataCache;
+import org.apache.helix.model.*;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * A rebalance tool that generate an resource partition assignment based on the input.
+ * Note the assignment won't be automatically applied to the cluster. Applications are supposed to
+ * apply the change.
+ */
```

Comment:
Good to have extensive javadoc on how to use this tool. Perhaps you  intend users to see the tests, in which case you can point to a test on how to use it. One example with zk and one without will be ideal. Thanks.

Reply:
Good suggestion. Will add it.

Reply:
I cannot emphasize the importance of an example more. Like I mentioned elsewhere, I would suggest adding a WeightBasedRebalanceExample class, in which you can provide example of a few different scenarios (number of partitions as weight, partitin size as weight, a combination of these two, etc.). Add plenty of javadoc in the example class to demonstrate the usage.

Reply:
I've already created the class, haha.

I will start with 3 examples, basic case, use ZK, and topology aware.